### PR TITLE
Make EmailSender a mock-able and testable API #6292

### LINKED ISSUE #6223: Refactor logic.automated package
The `logic.automated` package perhaps is the most undocumented and unmaintained section of the back-end portion of the application. Just to name a few of the existing problems, many of them serious:
- ~~No clear division between `WorkerServlet` and `AutomatedRemindersServlet`. They might as well be the same class. Analogously, the distinction between the terms "task queue", "cron jobs", and "automated actions" is vague especially for new developers.~~ Solved.
- ~~Servlets are directly accessible by the UI by navigating to the URL (e.g `/someServlet`); placing Servlets in logic is not really accurate. It should be in UI package.~~ Solved.
- ~~Too many duplications: each newly-defined servlet needs to be mapped in `web.xml`, `Const.AutomatedActionNames`, etc. Using template pattern (like how normal actions are designed) would alleviate this problem.~~ Solved.
- Not testable; servlets are by restriction not testable, and putting every operation in the servlet makes those operations not testable in return.

In the spirit of maintainability and also testability, this package is a very good candidate for refactoring.

- [x] #6224
- [x] #6233 
- [x] #6242 
- [x] #6243 
- [x] #6273 
- [x] #6302 
- [ ] #6292

### LINKED ISSUE #6292: Make EmailSender a mock-able and testable API
Part ~~six~~ seven and the last part of #6223.

This follows the idea of `TaskQueuer` and `MockTaskQueuer` introduced in #6275 and #6276, thus put on hold until those two PRs are merged.

### PR DESCRIPTION
**Pull request by wkurniawan07**
Fixes #6292 

~~Usual protocol of PR against another branch first :p~~

- First commit is self-explanatory.
- Next seven commits are the migrations. Some noteworthy pointers:
  - Task queue for student course registration invites are now extended to individual invites (not just course-wide invites), including after google id reset
  - Added new task queue for instructor course registration invites; this is so that student and instructor registration invites are treated similarly
  - For the above two pointers, new tests for the task queue worker actions are also added, in the style of #6308
- Afterwards comes the `MockEmailSender`, which is designed with heavy inspiration from `MockTaskQueuer`, and the new tests. There is no loss in details this time because the removed original tests are immediately replaced by newer ones.

And since this is the last part, this fixes #6223 at last. Indeed, all the items listed in that issue thread are now fixed.

### CODE REVIEW

**Comment by kanghj** [comment]
@JosephineKwa, I'm busy the next 2 weeks and won't be able to finish reviewing this. Can you review this one instead?

File: src/main/java/teammates/logic/api/EmailSender.java [hunk 1]
```diff
@@ -56,44 +63,27 @@ private void sendEmailCopyWithJavamail(EmailWrapper message) throws EmailSending
...
     }
     
-    private void logSevereForErrorInSendingItem(String itemType, EmailWrapper message, Exception e) {
-        log.severe("Error in sending " + itemType + ": " + (message == null ? "" : message.getInfoForLogging())
-                   + "\nCause: " + TeammatesException.toStringWithStackTrace(e));
+    /**
+     * Gets the emails sent.
+     * This method is used only for testing, where it is overridden.
+     * 
+     * @throws UnsupporedOperationException if used in production, where it is not meant to be
```
Comment (kanghj) [inline src/main/java/teammates/logic/api/EmailSender.java]: typo on `UnsupporedOperationException `

Comment (wkurniawan07) [inline src/main/java/teammates/logic/api/EmailSender.java]: Fixed, along with the ones in `TaskQueuer`.

**Comment by wkurniawan07** [comment]
@JosephineKwa it's green so should be ready for review :p

File: src/main/java/teammates/ui/controller/AdminInstructorAccountAddAction.java [hunk 2]
```diff
@@ -118,8 +121,17 @@ protected ActionResult execute() {
...
-        String joinLink = logic.sendJoinLinkToNewInstructor(instructorList.get(0), data.instructorShortName,
-                                                            data.instructorInstitution);
+        String joinLink = Config.getAppUrl(Const.ActionURIs.INSTRUCTOR_COURSE_JOIN)
+                                .withRegistrationKey(StringHelper.encrypt(instructorList.get(0).key))
+                                .withInstructorInstitution(data.instructorInstitution)
+                                .toAbsoluteString();
+        EmailWrapper email = new EmailGenerator().generateNewInstructorAccountJoinEmail(
+                instructorList.get(0).email, data.instructorShortName, joinLink);
+        try {
+            emailSender.sendEmail(email);
```
Comment (JosephineKwa) [inline src/main/java/teammates/ui/controller/AdminInstructorAccountAddAction.java]: Hey @wkurniawan07, I was just wondering if there was any guideline as to when we send the email through `sendEmail`, and when we should schedule the sending of emails. Would there be an issue if we send the email directly here even if there are many send email tasks in the queues waiting to be executed?

Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/AdminInstructorAccountAddAction.java]: Whatever's in the task queue has nothing to do with this line.

Comment (JosephineKwa) [inline src/main/java/teammates/ui/controller/AdminInstructorAccountAddAction.java]: Alright, should be good to merge then

**Pull request merged**